src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java

```diff
@@ -419,36 +420,24 @@ private static DetailAST getFirstNonArrayDeclaratorParent(DetailAST ast) {
      * @return type node.
      */
     private static DetailAST getTypeLastNode(DetailAST ast) {
-        DetailAST result = ast.findFirstToken(TokenTypes.TYPE_ARGUMENTS);
-        if (result == null) {
-            result = getIdentLastToken(ast);
-            if (result == null) {
-                result = getArrayType(ast);
-            }
+        final DetailAST typeLastNode;
+        final DetailAST parent = ast.getParent();
+        final boolean isPrecededByTypeArgs =
+                parent.findFirstToken(TokenTypes.TYPE_ARGUMENTS) != null;
+        final Optional<DetailAST> objectArrayType = Optional.ofNullable(getIdentLastToken(ast));
+
+        if (isPrecededByTypeArgs) {
+            typeLastNode = parent.findFirstToken(TokenTypes.TYPE_ARGUMENTS)
+                    .findFirstToken(TokenTypes.GENERIC_END);
+        }
+        else if (objectArrayType.isPresent()) {
+            typeLastNode = objectArrayType.get();
         }
         else {
-            result = result.findFirstToken(TokenTypes.GENERIC_END);
+            typeLastNode = parent.getFirstChild();
         }
-        return result;
-    }
 
-    /**
-     * Ascends through array type declarations to get type,
-     * e.g. 'char' literal from 'char [][]'.
-     *
-     * @param ast the method or type parameter to find type of
-     * @return type of parameter
-     */
-    private static DetailAST getArrayType(DetailAST ast) {
-        DetailAST typeAst = ast.getFirstChild();
-        if (TokenUtil.isOfType(typeAst, TokenTypes.RBRACK, TokenTypes.ANNOTATIONS)) {
-            while (!TokenUtil.isOfType(typeAst, TokenTypes.TYPE_ARGUMENT, TokenTypes.TYPE)
-                    && typeAst.getParent() != null) {
-                typeAst = typeAst.getParent();
-            }
-            typeAst = typeAst.getFirstChild();
-        }
-        return typeAst;
```

Comment:
Ascending through ARRAY_DECLARATOR, ANNOTATIONS, and TYPE_ARGUMENT tokens isn't necessary anymore; all these elements are sequential under the new TYPE ast. See https://github.com/checkstyle/checkstyle/pull/10104/files#r652088410